pkg/config/defaults_test.go

```diff
@@ -0,0 +1,51 @@
+package config
+
+import (
+	"io"
+	"io/ioutil"
+	"strings"
+	"testing"
+)
+
+func TestReadPipelineDefaults(t *testing.T) {
+
+	var d PipelineDefaults
+
+	t.Run("Success case", func(t *testing.T) {
+		d0 := strings.NewReader("general:\n  testStepKey1: testStepValue1")
+		d1 := strings.NewReader("general:\n  testStepKey2: testStepValue2")
+		err := d.ReadPipelineDefaults([]io.ReadCloser{ioutil.NopCloser(d0), ioutil.NopCloser(d1)})
+
+		if err != nil {
+			t.Errorf("Got error although no error expected: %v", err)
+		}
+
+		t.Run("Defaults 0", func(t *testing.T) {
+			if d.Defaults[0].General["testStepKey1"] != "testStepValue1" {
+				t.Errorf("got: %v, expected: %v", d.Defaults[0].General["testStepKey1"], "testStepValue1")
+			}
+		})
+
+		t.Run("Defaults 1", func(t *testing.T) {
+			if d.Defaults[1].General["testStepKey2"] != "testStepValue2" {
```

Comment:
As far as I understand the code we have the various default configurations layered one above the other. This we do not have in the groovy piper lib. In general I appreciate having that. But: what we finally need is: we must be able to ask for the _merged_ defaults. Hence I think the test should work without the index in the `d.Defaults`. 

More technically:

Given this:

```
d0 := strings.NewReader("general:\n  testStepKey1: testStepValue1\n  testStepKey3: base")
d1 := strings.NewReader("general:\n  testStepKey2: testStepValue2\n  testStepKey3: overwritten")
err := d.ReadPipelineDefaults([]io.ReadCloser{ioutil.NopCloser(d0), ioutil.NopCloser(d1)})
```

What we get currently is:

```
[
    {map[testStepKey1:testStepValue1 testStepKey3:base] map[] map[]}
    {map[testStepKey2:testStepValue2 testStepKey3:overwritten] map[] map[]}
]

```

What I would expect is:

```

    {map[testStepKey1:testStepValue1  testStepKey2:testStepValue2 testStepKey3:overwritten] map[] map[]
```

Reply:
ReadPipelineDefaults should not merge it yet but only read the defaults. That is intended.

Merging will then happen in the respective step context using GetStepConfig

Reply:
... aaahhhh, thats OK for me. This is an improvement compared to the groovy approach. Here we can see from which config layer a value comes from, in groovy we have only the merged total view.